Add model builder overview doc (#12557)
dotnet/docs: docs/machine-learning/automate-training-with-model-builder.md

Model Builder uses automated machine learning (AutoML) to evaluate different models. It chooses the best model for your scenario, without any manual tuning. Once the optimal model is selected, Model Builder generates code to add the model to your application.

> [!NOTE]
> Model Builder is currently in Preview.

Model Builder is a graphical Visual Studio extension. 

![Model Builder Visual Studio extension user interface animation](media/ml-dotnet-model-builder.gif)

## Scenarios

You can bring many different scenarios to Model Builder, to generate a machine learning model for your application.

A scenario is a description of the type of prediction you want to make on your data. For example: predict future product sales volume based on historical sales data.

In Model Builder, you need to map your scenario onto an **ML.NET task**. Model Builder currently supports **regression** (predict numbers) and **classification** (predict categories) tasks.

Here are some example scenarios, with the corresponding task.

> |||

Author: Done (or removed tables)

|-|-|
|Forecast sales|regression|
|Predict prices|regression|
|Predict time to failure|regression|
|Detect email spam|classification|
|Work out whether customer feedback is good or bad|classification|
|Detect fraudulent credit card transactions|classification|

## Data

Once you have mapped your scenario onto a task, Model Builder asks you to provide a training dataset. The data is used to train, evaluate, and choose the best model for your scenario. A training dataset is a table of rows of training examples, and columns of attributes. Each row has:
- a **label** (the attribute that you want to predict)
- **features** (attributes that are used as inputs to predict the label).

For the house-price prediction scenario, the features could be:
- the square footage of the house
- the number of bedrooms and bathrooms

> - the zip code.

Author: Fixed

|-|-|-|
|[taxi fare prediction](https://github.com/dotnet/machinelearning-samples/blob/master/datasets/taxi-fare-train.csv)|Fare|Trip time, distance|
|[product sales forecasting](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/AnomalyDetection_Sales/SpikeDetection/Data/product-sales.csv)|Product Sales|Month|
|[website comment sentiment analysis](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/BinaryClassification_SentimentAnalysis/SentimentAnalysis/Data/wikiDetoxAnnotated40kRows.tsv)|Label (0 when negative sentiment, 1 when positive)|Comment, Year|
|[credit card fraud detection](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/BinaryClassification_CreditCardFraudDetection/CreditCardFraudDetection.Trainer/assets/input/creditcardfraud-dataset.zip)|Amount, V1-V28 (anonymized features|Class (1 when fraudulent, 0 otherwise)|
|[issue classification](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/end-to-end-apps/MulticlassClassification-GitHubLabeler/GitHubLabeler/Data/corefx-issues-train.tsv)|Area|Title, Description|

> Model Builder currently places the following limitations on the data that it imports:

Author: Moved into its own section

* SQL Server has a limit of 100-K rows for training
* Microsoft SQL Server Data Tools for Visual Studio 2017 isn't supported.

## Train

Once you have provided the task and the data, Model Builder trains the model.

Training in Model Builder is the process of trying different algorithms and settings, and evaluating how accurately they perform.

You can provide a training time. In general, training for a longer time produces a more accurate model. More training time is also required as the size of the training dataset increases. The following table gives some training time guidelines for datasets of increasing size.

Dataset Size  | Dataset Type       | Avg. Time to train
------------- | ------------------ | --------------
0 - 10 Mb     | Numeric and Text   | 10 sec
10 - 100 Mb   | Numeric and Text   | 10 min 
100 - 500 Mb  | Numeric and Text   | 30 min 
500 - 1 Gb    | Numeric and Text   | 60 min 
1 Gb+         | Numeric and Text   | 3 hour+ 

## Evaluate

Model Builder by splits the training data into a training set and a test set. The training data (80%) is used to train your model and the test data (20%) is used to evaluate your model. 

### Regression (predicting numbers)

> The default metric for regression problems is **RSquared**. 1 is the best possible value. The closer **RSquared** is to 1, the better your model is.

Reviewer: Consistently bolden

```suggestion
The default metric for regression problems is **RSquared**. 1 is the best possible value. The **closer RSquared is to 1, the better your model is**.
```

Author: Fixed.

Other metrics reported, such as absolute-loss, squared-loss, and RMS loss can be used to understand your model, and compare it with other regression models. 

### Classification (predicting categories)

The default metric for classification problems is **accuracy**. Accuracy defines the proportion of correct predictions your model makes over the test dataset. The **closer to 100%, the better it is**. 

> Other metrics reported such as AUC (Area under the curve), which measures the true positive rate vs. the false positive rate should be greater than 0.50, for models to be acceptable. 

Reviewer: Suggestion

```suggestion
Other metrics reported such as AUC (Area under the curve), which measures the true positive rate vs. the false positive rate, should be greater than 0.50 for models to be acceptable. 
```

Additional metrics such as F1 score can be used to control the balance between precision (ratio of correct predictions to the total predictions of that class) and recall (proportion of correct predictions to the total actual members of that class).

> When there are more than two categories, are two types of accuracy:

Reviewer: Suggestion

```suggestion
When there are more than two categories, there are two types of accuracy:
```

Author: Section has been refactored

---
title: What is Model Builder and how does it work?
description: How to use the ML.NET Model Builder to automatically train a machine learning model
author: natke
ms.date: 05/29/2019
ms.custom: overview
#Customer intent: As a developer, I want to use Model Builder to automatically train a model using a visual interface. 
---
# What is Model Builder & how does it work?

> Model Builder is visual interface to automatically train custom ML.NET machine learning models.  You don't need machine learning expertise to use Model Builder. All you need is some data, and a problem to solve. Model Builder also generates the code to add the model your .NET application.

Author: uses

> Model Builder is visual interface to automatically train custom ML.NET machine learning models.  You don't need machine learning expertise to use Model Builder. All you need is some data, and a problem to solve. Model Builder also generates the code to add the model your .NET application.

Author: Used a combination of your text and mine:

ML.NET Model Builder is an easy-to-understand visual interface to build, train, and deploy custom machine learning models. 

Model Builder uses automated machine learning (AutoML) to explore different machine learning algorithms and settings to help you find the one that best suits your scenario.

You don't need machine learning expertise to use Model Builder. All you need is some data, and a problem to solve. Model Builder generates the code to add the model your .NET application.

Model Builder uses automated machine learning (AutoML) to evaluate different models. It chooses the best model for your scenario, without any manual tuning.

Model Builder is a graphical Visual Studio extension. 

> [!NOTE]
> Model Builder is currently in Preview.

![Model Builder Visual Studio extension user interface animation](media/ml-dotnet-model-builder.gif)

## Scenarios

You can bring many different scenarios to Model Builder, to generate a machine learning model for your application.

A scenario is a description of the type of prediction you want to make on your data. For example: predict future product sales volume based on historical sales data.

> In Model Builder, you need to map your scenario onto an [ML.NET task](resources/tasks.md). Model Builder currently supports **regression** (predict numbers) and **classification** (predict categories) tasks.

Author: Or ... you can use Model Builder for ...

Binary classification is used to categorize data into two classes (yes/no; pass/fail; true/false; positive/negative). It can be used to answer questions such as:

- Is this email spam?
- Which applicants may be eligible for membership?
- Which accounts may not pay their invoices on time?
- Is this credit card transaction fraudulent?

If your scenario requires classification into two categories, you can use this template.

#### Issue classification (multiclass classification)

Issue classification can be used to categorize customer feedback issues using the issue title and description. It is an example of the multi-class classification task.

Multiclass classification can be used to categorize data into three or more classes. It can be used to answer questions such as:

- To which department should I route a support ticket?
- What category does a financial transaction belong in?
- What is the priority of a customer issue?

You can use the issue classification template for your scenario if you want to categorize data into three or more categories.

#### Price prediction (regression)

> Price prediction can be used to predict house prices using location, size, and features of the house, such as number of bedrooms and bathrooms. It is an example of the regression task.

Author: Price prediction can be used to predict house prices using location, size, and other characteristics of the house.

### Choose the output to predict (label)

A dataset is a table of rows of training examples, and columns of attributes. Each row has:
- a **label** (the attribute that you want to predict)
- **features** (attributes that are used as inputs to predict the label).

For the house-price prediction scenario, the features could be:
- the square footage of the house
- the number of bedrooms and bathrooms
- the zip code

The label is the historical house price for that row of square footage, bedroom, and bathroom values and zip code.

![Table showing rows and columns of house price data with features consisting of size rooms zip code and price label](media/model-builder-data.png)

### Data formats

Model Builder places the following limitations on the data:

- Data must be stored in a file (.csv or .tsv with a header row), or in a SQL server database.
- There is a limit of 1 GB on the training dataset
- SQL server has a limit of 100,000 rows for training
- Data from SQL server is copied from the server to your local machine before training

> - Microsoft SQL Server Data Tools for Visual Studio 2017 is not supported

Author: Because this is not a limitation of the data

### Example datasets

If you don't have your own data yet, try out one of these datasets:

> |Data|Label|Features|

Author: Reinforce the Scenario/Task learning here

|-|-|-|
|[taxi fare prediction](https://github.com/dotnet/machinelearning-samples/blob/master/datasets/taxi-fare-train.csv)|Fare|Trip time, distance|
|[product sales forecasting](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/AnomalyDetection_Sales/SpikeDetection/Data/product-sales.csv)|Product Sales|Month|
|[website comment sentiment analysis](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/BinaryClassification_SentimentAnalysis/SentimentAnalysis/Data/wikiDetoxAnnotated40kRows.tsv)|Label (0 when negative sentiment, 1 when positive)|Comment, Year|

> |[credit card fraud detection](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/BinaryClassification_CreditCardFraudDetection/CreditCardFraudDetection.Trainer/assets/input/creditcardfraud-dataset.zip)|Amount, V1-V28 (anonymized features)|Class (1 when fraudulent, 0 otherwise)|

Author: These are around the wrong way

Model Builder uses automated machine learning (AutoML), which automatically explores different machine learning algorithms and settings to find the one that best suits your scenario. Model Builder does not require any input or tuning from you.

### How long should I train for?

You can provide a training time. In general, training for a longer time produces a more accurate model. More training time is also required as the size of the training dataset increases. The following table gives some training time guidelines for datasets of increasing size.

Dataset Size  | Dataset Type       | Avg. Time to train
------------- | ------------------ | --------------
0 - 10 Mb     | Numeric and Text   | 10 sec
10 - 100 Mb   | Numeric and Text   | 10 min 
100 - 500 Mb  | Numeric and Text   | 30 min 
500 - 1 Gb    | Numeric and Text   | 60 min 
1 Gb+         | Numeric and Text   | 3 hour+ 

The exact time to train also depends on:

- the type of columns that is, text vs numeric
- the type of machine learning task (regression or classification)
- the number of rows used for training
- the number of feature columns used for training

## Evaluate

> Model Builder splits the training data into a training set and a test set. The training data (80%) is used to train your model and the test data (20%) is held back to evaluate your model. Evaluation is the process of using the trained model to make predictions with the new test data, and then measuring how good the predictions are. The metrics used for evaluation depend on the ML task. For more information, see [model evaluation metrics](resources/metrics.md).  

Author: Switch out "Evaluation is ..." sentence to go first

---
title: What is Model Builder and how does it work?
description: How to use the ML.NET Model Builder to automatically train a machine learning model
author: natke
ms.date: 06/26/2019
ms.custom: overview
#Customer intent: As a developer, I want to use Model Builder to automatically train a model using a visual interface. 
---

> # What is Model Builder & how does it work?

Reviewer 2: ```suggestion
# What is Model Builder and how does it work?
```

> ML.NET Model Builder is an easy-to-understand visual interface to build, train, and deploy custom machine learning models. 

Reviewer 2: ```suggestion
ML.NET Model Builder is an easy-to-understand graphical Visual Studio extension to build, train, and deploy custom machine learning models. 
```

Model Builder uses automated machine learning (AutoML) to explore different machine learning algorithms and settings to help you find the one that best suits your scenario.

> You don't need machine learning expertise to use Model Builder. All you need is some data, and a problem to solve. Model Builder generates the code to add the model your .NET application.

Reviewer 2: ```suggestion
You don't need machine learning expertise to use Model Builder. All you need is some data, and a problem to solve. Model Builder generates the code to add the model to your .NET application.
```

> Model Builder is a graphical Visual Studio extension. 

Reviewer 2: This can be removed as it was combined above.

![Model Builder Visual Studio extension user interface animation](media/ml-dotnet-model-builder.gif)

> [!NOTE]
> Model Builder is currently in Preview.

## Scenarios

You can bring many different scenarios to Model Builder, to generate a machine learning model for your application.

> A scenario is a description of the type of prediction you want to make on your data. For example

Reviewer 2: ```suggestion
A scenario is a description of the type of prediction you want to make on your data. For example:
```

- Is this credit card transaction fraudulent? (fraud detection)

If your scenario requires classification into two categories, you can use this template with your own dataset.
 
#### Issue classification (multiclass classification)

Issue classification can be used to categorize customer feedback (for example, on GitHub) issues using the issue title and description. It is an example of the multi-class classification task.

Multiclass classification can be used to categorize data into three or more classes. It can be used to answer questions such as:

- To which department should I route a support ticket? (support ticket routing)
- What is the priority of a customer issue? (customer issue prioritization)
- What category does a product belong to? (product classification)
- What type of document? (document/email classification)

You can use the issue classification template for your scenario if you want to categorize data into three or more categories.

#### Price prediction (regression)

Price prediction can be used to predict house prices using location, size, and other characteristics of the house. It is an example of the regression task.

> Price prediction can be used to predict house prices using location, size, 

Reviewer 2: ```suggestion
Price prediction can be used to predict house prices using location and size.
```

Author: That was actually an accidental repetition. Well spotted

Regression is used to predict numbers. It can be used to answer questions such as:

- What price will a house sell for? (price prediction)
- After how much time will a mechanical part require maintenance? (predictive maintenance)
- What is the moisture content in this dryer? (machine monitoring)
- What will the total annual sales for this region be? (sales forecasting)

> You can use the price prediction template for your scenario if you want to predict a numerical value with your own dataset

Reviewer 2: ```suggestion
You can use the price prediction template for your scenario if you want to predict a numerical value with your own dataset.
```

|Fraud detection|binary classification|[credit card data](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/getting-started/BinaryClassification_CreditCardFraudDetection/CreditCardFraudDetection.Trainer/assets/input/creditcardfraud-dataset.zip)|Class (1 when fraudulent, 0 otherwise)|Amount, V1-V28 (anonymized features)|
|Text classification|multiclass classification|[GitHub issue data](https://github.com/dotnet/machinelearning-samples/blob/master/samples/csharp/end-to-end-apps/MulticlassClassification-GitHubLabeler/GitHubLabeler/Data/corefx-issues-train.tsv)|Area|Title, Description|

## Train

Once you select your scenario, data, and label, Model Builder trains the model.

### What is training?

Training is an automatic process by which Model Builder teaches your model how to answer questions for your scenario. Once trained, your model can make predictions with input data that it has not seen before. For example, if you are predicting house prices and a new house comes on the market, you can predict its sale price.

> Model Builder uses automated machine learning (AutoML), which automatically explores different machine learning algorithms and settings to find the one that best suits your scenario. Model Builder does not require any input or tuning from you. No prior expertise for ML is required.

Reviewer 2: This seems redundant as it was stated previously and could be removed for brevity.

------------- | ------------------ | --------------
0 - 10 Mb     | Numeric and Text   | 10 sec
10 - 100 Mb   | Numeric and Text   | 10 min 
100 - 500 Mb  | Numeric and Text   | 30 min 
500 - 1 Gb    | Numeric and Text   | 60 min 
1 Gb+         | Numeric and Text   | 3 hour+ 

The exact time to train also depends on:

- the type of columns that is, text vs numeric
- the type of machine learning task (regression or classification)
- the number of rows used for training
- the number of feature columns used for training

Model Builder has been tested for scale with a 1-TB dataset, but building a high-quality model for that size of dataset can take up to four days!

## Evaluate

Evaluation is the process of using the trained model to make predictions with new test data, and then measuring how good the predictions are.

Model Builder splits the training data into a training set and a test set. The training data (80%) is used to train your model and the test data (20%) is held back to evaluate your model.  The metrics used for evaluation depend on the ML task. For more information, see [model evaluation metrics](resources/metrics.md).  

### Sentiment analysis (binary classification)

> 

Reviewer 2: Could you link to the sections in the metrics doc instead of reproing here?

Author: Possibly. For expediency, I'll leave it for now, and make a note for later

> The default metric for binary classification problems is **accuracy**. Accuracy defines the proportion of correct predictions your model makes over the test dataset. The **closer to 100%, the better it is**. 

Reviewer 2: I'd move all the default metrics to either the top or bottom of the each section for consistency and readability.
